### PR TITLE
Annotation WIP

### PR DESCRIPTION
**Pull request by andrewkern**
Hey all-

I've sketched out what i think are the main bits of the infrastructure for dealing with annotation. I've aimed to mirror what we do with genetic maps, although I figure if we do this right we don't have to host the data at all, and instead we can rely on Ensembl (or similar) as I'm doing now. 

The basic idea is that annotations will be downloaded to the cache as a GFF file. Annotations will have citations, ids, etc, just as do genetic_maps. Handling of the actual annotation files is currently done using [pybedtools](https://daler.github.io/pybedtools/), although I haven't build anything in other than a function that gets the annotations from a chromosome (we will need to make this useful with SLiM-- tagging @mufernando to help me).

Currently I only have annotations defined for HomSap. I will build this out further for the rest of the catalog once people have a look over this code. 

thanks.

### CODE REVIEW

**Comment by ndukler** [comment]
We should probably have some way of handling the sex chromosomes as part of this. Dunno if it's at the species level or the assembly level though.

**Comment by andrewkern** [comment]
do you mean label sex chroms? they are in the GFFs

**Comment by andrewkern** [comment]
ah got it. i don't *believe* that should hang out here though. i imagine we'd want those decorations in the Chromosome or Contig class

**Comment by ndukler** [comment]
Sure, we're going to need to setup some logic to pass those along when we retrieve and parse the assembly.

**Comment by andrewkern** [comment]
ugh i see... there is no `pybedtools` package on conda for windows... what do i do in this case?

**Comment by mufernando** [comment]
Thanks for this, @andrewkern!

It looks good overall, but I have one suggestion:  I think we should make the class more general than `GTF` and instead call it `Annotation`. In principle, we should be able handle any `pybedtools` supported format, such as BED, GFF, GTF.

I also agree with @ndukler that this should ultimately not be a property of `species`, but of a assembly. For now, though, it's fine to stay where it's at.

Now, to start connecting this with the SLiM side we need to implement the following:

- `GenomicElement`
    - Intervals
    - Mutation types (and weights)

- Something more meta that houses all genomic elements which will handle the non-overlappingness of the elements. Further, deal with the neutral mutations and the fact that there has to be an element at the end of the chromosome (`chromosome_length-1`), so that SLiM actively records the right length.

- A mapping of Annotation objects to Elements. This gets tricky and might determine our support for GFF files vs. BED. GFF files have a feature column that could each be a GenomicElement. BED files on the other hand are usually made on a per-feature basis.

(btw. should I move this to a separate issue so we can discuss in more detail?)

**Comment by andrewkern** [comment]
sure yeah happy to put this in a separate issue. The GTF() class is poorly named now, but this is the rough outline of how we will get annotations *in*

**Comment by mufernando** [comment]
> ugh i see... there is no `pybedtools` package on conda for windows... what do i do in this case?

It seems for now you'll have to skip those tests in windows as they did in #397

**Comment by grahamgower** [comment]
Hey @andrewkern, I don't about motivation to support cygwin-less windows. However, it seems that pybedtools states "windows is not supported" http://daler.github.io/pybedtools/main.html. Possibly this is because pybedtools depends upon bedtools and pysam. bedtools says nothing about windows on its installation page (https://bedtools.readthedocs.io/en/latest/content/installation.html), and there's this open ticket about pysam windows failures: https://github.com/pysam-developers/pysam/issues/575.

So what is the motivation for including pybedtools as a dependency? What parts of the GFF file are we expecting to use? Are we anticipating a need to do set operations on intervals? I vote we avoid that kind of complexity if possible. GFF is just tab separated data with a header, so it should be possible to extract the basics with `numpy.loadtxt()`, given an appropriate dtype.

**Comment by mufernando** [comment]
We can swap pybedtools to [pyranges](https://github.com/biocore-ntnu/pyranges). It seems to support windows, [parses](https://pyranges.readthedocs.io/en/latest/autoapi/pyranges/readers/index.html) BED/GFF/GTF and does interval operations even faster than pybedtools.

**Comment by andrewkern** [comment]
> So what is the motivation for including pybedtools as a dependency? What parts of the GFF file are we expecting to use? Are we anticipating a need to do set operations on intervals? I vote we avoid that kind of complexity if possible. GFF is just tab separated data with a header, so it should be possible to extract the basics with `numpy.loadtxt()`, given an appropriate dtype.

Yeah i can definitely avoid it, but was thinking we would want something to do interval operations with (e.g., to deal with overlapping annotations or quick filtering). I can easily just use numpy or pandas even. I'll look in to the pyranges thing.

@grahamgower how does the rest of this look to you?

File: stdpopsim/annotations.py
Comment (grahamgower) [inline stdpopsim/annotations.py:20]: I might have missed it, but I didn't see this `cd` function used anywhere.

File: stdpopsim/annotations.py
Comment (grahamgower) [inline stdpopsim/annotations.py:120]: Probably want to just trim from the start of the `id` string. E.g.
```
if id.startswith("chr"):
    id = id[len("chr"):]
```

But shouldn't the chomosome names match those used along with the species/assembly definition? I guess the Assembly class needs to have a `chr_prefix` attribute, which says if the chromosome names have a prefix, so we know when to trim it and when not to...

File: stdpopsim/annotations.py
Comment (grahamgower) [inline stdpopsim/annotations.py:33]: I second @mufernando's motion that this class be named `Annotation`.

File: tests/test_annotations.py
Comment (grahamgower) [inline tests/test_annotations.py:71]: The comment and docstring here seem misplaced or out of date?

File: stdpopsim/annotations.py
Comment (grahamgower) [inline stdpopsim/annotations.py:79]: It might be nice to factor out the duplicated download/cache code between this class and the GeneticMap class.

File: tests/test_annotations.py
Comment (grahamgower) [inline tests/test_annotations.py:164]: 'TestAllGeneticMaps'... It looks like this is not related to annotations or the GTF class?

**Comment by grahamgower** [comment]
> Yeah i can definitely avoid it, but was thinking we would want something to do interval operations with (e.g., to deal with overlapping annotations or quick filtering). I can easily just use numpy or pandas even. I'll look in to the pyranges thing.

I spent a few minutes playing with some annotations just now and can see the advantages. But I think we need to consider what annotations we're likely to use---if only to reduce the load times. I did some timing tests, just to load the (uncompressed) `Homo_sapiens.GRCh37.87.gff3` file (https://gist.github.com/grahamgower/edab09eb43fbfcf4a4c92c0606940039).

```
$ ./time_gff_loading.py Homo_sapiens.GRCh37.87.gff3
pybedtools_BedTool took 8 seconds
pyranges_read_gff3 took 30 seconds
load_gff_numpy took 18 seconds
load_gff_python took 3 seconds
```

**Comment by andrewkern** [comment]
yeah we will want to stick with GFF files from Ensembl I think and we will want to keep them compressed. I wonder if we can use a pandas reader to speed this up further?

**Comment by mufernando** [comment]
pyranges is [using pandas](https://github.com/biocore-ntnu/pyranges/blob/master/pyranges/readers.py) I think.

**Comment by andrewkern** [comment]
well that's how i would have done it too. shit. okay i can implement a python implementation like @grahamgower has on that gist

**Comment by mufernando** [comment]
Just tested [scikit-allel's GFF3 reader](https://scikit-allel.readthedocs.io/en/stable/io.html#gff3) and it's almost as fast as @grahamgower's pure Python version. With a bunch of added benefits (e.g., being able to extract chunks). See the time testing [here](https://gist.github.com/mufernando/750c45aec17d146036c7a27122595214).

```
pybedtools_BedTool took 21 seconds
pyranges_read_gff3 took 40 seconds
load_gff_numpy took 29 seconds
load_gff_python took 6 seconds
allel_gff3 took 11 seconds
```

**Comment by andrewkern** [comment]
happy to try it out! will reimplement that bit

File: stdpopsim/annotations.py
Comment (andrewkern) [reply stdpopsim/annotations.py:79]: the directory structure here is different enough that it would be a pain i think. little reward vs the bit of duplicated code IMO

**Comment by grahamgower** [comment]
Hey @andrewkern, I'm still rather concerned about how slow it is to load up this 2.6 million line gff file. Presumably this is most problematic for HomSap, as lots of good annotations exist. Given that the download step is going to take a while and is a one-off, perhaps we could do some preprocessesing on the gff when the file gets downloaded into the cache? One obvious choice we might make is to split the gff by chromosome. But we could also split by the source or feature columns. (E.g. it would be possible to have it cached like `cache_dir/annotations/Homo_sapiens.GRCh37.87/1/ensembl/exon.gff.gz`)

Sources:
```
$ awk '!/^#/ {print $2}' Homo_sapiens.GRCh37.87.gff3 | sort  |
uniq -c
  22397 cpg
 360236 ensembl
 696421 ensembl_havana
  61761 Eponine
  87144 firstef
     84 GRCh37
1397884 havana
    124 insdc
    626 tRNAscan-SE
```
I don't know anything about gene annotations, so it's not clear to me whether we'd favour using `ensembl`, `ensembl_havana,` or `havana`. But I can't see an obvious use case for loading more than one at a time.

Feature/type
```
$ awk '!/^#/ {print $3}' Homo_sapiens.GRCh37.87.gff3 | sort  |
uniq -c
  25944 aberrant_processed_transcript
 171928 biological_region
 724218 CDS
     28 C_gene_segment
     25 chromosome
1195764 exon
 148015 five_prime_UTR
  26998 gene
    148 J_gene_segment
  11773 lincRNA
   7109 lincRNA_gene
   3174 miRNA
   3111 miRNA_gene
  81787 mRNA
     22 mt_gene
     25 nc_primary_transcript
  13046 NMD_transcript_variant
  11060 processed_pseudogene
  28551 processed_transcript
  17217 pseudogene
    858 pseudogenic_transcript
   2038 RNA
    537 rRNA
    533 rRNA_gene
   1531 snoRNA
   1459 snoRNA_gene
   1930 snRNA
   1923 snRNA_gene
     59 supercontig
 132262 three_prime_UTR
  13373 transcript
     37 VD_gene_segment
    194 V_gene_segment
```

**Comment by grahamgower** [comment]
Oh, it seems there are chromosome-specific files already available. Can we start by using those?
http://ftp.ensembl.org/pub/grch37/release-87/gff3/homo_sapiens/

**Comment by jeromekelleher** [comment]
I agree with @grahamgower here that we should use the cache to avoid some computation here. I have another suggestion to throw into the mix though: what if we (shock horror) didn't use text to store this stuff but used [zarr](https://zarr.readthedocs.io/en/stable/) instead? Then we're not quite so performance sensitive in the GFF parsing, and the client code is simplified by being able to work with numpy arrays rather than some upstream library's API.

File: stdpopsim/annotations.py
Comment (jeromekelleher) [inline stdpopsim/annotations.py:63]: Not sure this logic holds up for ``is_cached``

File: stdpopsim/annotations.py
Comment (jeromekelleher) [inline stdpopsim/annotations.py:103]: Why the specific check for None here? Seems like lots of other ways to have a bad id. I'd just remove the check.

**Comment by grahamgower** [comment]
> think we should look into (a) retrieving stuff programatically from Ensembl first and then (b) storing the data in the cache as zarr or something. I'd like to avoid "classical" bioinfo data formats and the messy dependencies if at all possible.

Regards (a), it would seem that enseml don't provide access to resources for non-current assemblies. E.g. human reference GRCh37.p13 should be here but is not:
https://rest.ensembl.org/info/genomes/assembly/GCA_000001405.14
Am I missing something?

Caching using zarr sounds nice though!

**Comment by andrewkern** [comment]
hey all-- zarr and preprocessing seems good to me [i quite like the zarr format]. let's talk about this on the call

**Comment by andrewkern** [comment]
I will heavily edit this PR after we merge #563, although one question I have is can we easily pull genome annotation via the REST API? 

I see a lot of REST endpoints that are locus specific, rather than genomewide, but presumably this can be done?

**Comment by jeromekelleher** [comment]
> I see a lot of REST endpoints that are locus specific, rather than genomewide, but presumably this can be done?

Good question. It might be a struggle to pipe that much data through JSON though - I think we want to convert to zarr now? So, we might add the script for converting GFF into zarr to our maintenance directory, and just make sure we're using the same version of Ensembl?

Although, if we *could* do it via the REST API that would be ideal.

**Comment by andrewkern** [comment]
yeah after looking further I'm pretty sure that we will need to download the GFF and convert to zarr rather than do this through the REST API. I could add the GFF to zarr script to this PR once we merge #563. what do you think Jerome?

**Comment by andrewkern** [comment]
hey @jeromekelleher check out this draft of the maintenance script for doing the download and conversion to zarr https://github.com/popsim-consortium/stdpopsim/pull/560/commits/55718c03710887ccbe0f2873464e6d58c70078c3

**Comment by andrewkern** [comment]
it's still rough, but wanted to check if this is what you have in mind. this will go through all the species, download the indicated annotation, and then shove them in a zarr group.

I still need to figure out how then to upload these automatically to AWS. do we have a way to do that?

**Comment by jeromekelleher** [comment]
Yep, looks ideal. We can tidy up the rough edges afterwards.

I'm sure there must be some way to automatically upload to AWS - you and @ndukler were the people dealing with this stuff though. How about you loop me in to whatever details you have and I'll see what we need to do?

**Comment by andrewkern** [comment]
well circleci is failing to build the scikit-allel wheel for some reason (anyone know why?) but i think the code is in good enough shape for review. 

this code deals with the maintenance and distribution of annotations. GFF files are downloaded by a maintenance script and then coverted to zarr files and pushed to AWS. the stdpopsim client then automatically downloads / caches the zarr locally, and the user can interface with the GFF annotations via an pandas dataframe that is produced upon request.

**Comment by grahamgower** [comment]
Cool! Thanks @andrewkern. It looks like you solved the CI issues here. I'll take a close look in the next few days.

**Comment by andrewkern** [comment]
do to from here includes:
- writing a function to filter for DNA element type (e.g., 'gene'). this is trivial given that the output is a pandas dataframe
- writing a function that goes from the filtered df to the mutation type inputs through the SLiM interface

**Comment by mufernando** [comment]
I assigned myself to review this so I can get started on linking this to SLiM. Hope to get to it by early next week at the most!

**Comment by andrewkern** [comment]
> do to from here includes:
> 
> * writing a function to filter for DNA element type (e.g., 'gene'). this is trivial given that the output is a pandas dataframe

i'm going to go ahead and write this and commit more edits to this PR

File: stdpopsim/annotations.py
Comment (grahamgower) [inline stdpopsim/annotations.py:140]: It isn't obvious to me how this case can happen. But I think it would be better if this raised a ValueError, instead of having a warning and returning None.

File: stdpopsim/annotations.py
Comment (grahamgower) [inline stdpopsim/annotations.py:119]: I suppose this will be the main entry point for consumers of annotations. So we'll need to have complete docstrings for this (and the other user-facing methods in this class). Happy for this to be punted down the road though. Maybe open a ticket tracking this once it's merged?

File: stdpopsim/catalog/HomSap/__init__.py
Comment (grahamgower) [inline stdpopsim/catalog/HomSap/__init__.py:232]: Ensembl seem to have a publication every year. How will we choose which one to cite? (I guess you chose this one because it corresponds with release 87?)

File: stdpopsim/catalog/HomSap/__init__.py
Comment (grahamgower) [inline stdpopsim/catalog/HomSap/__init__.py:224]: HomSap is now on GRCh38, so these annotations won't match the genome assembly in the catalog! Is there a way to make the version automagically match up?

File: stdpopsim/species.py
Comment (grahamgower) [inline stdpopsim/species.py:255]: docstring needed

File: tests/test_annotations.py
Comment (grahamgower) [inline tests/test_annotations.py:106]: s/genetic map/annotation/

File: tests/test_annotations.py
Comment (grahamgower) [inline tests/test_annotations.py:129]: s/genetic maps/annotations/

**Review by grahamgower — CHANGES_REQUESTED** [review]
Hey @andrewkern, I've gone through this fairly thoroughly now. It looks great, and I think it's functionally complete now! There are a few rough edges that can be easily fixed up, and others that can be punted down the road. The only blocker I see here is that you've included HomSap annotations for GRCh37, but we now have chromosome lengths/units for HomSap GRCh38.

Also, it would be good if you could run `flake8 --max-line-length 89 maintenance/anno*.py` --- this is currently missing from the circle CI checks, which is just an oversight.

File: maintenance/annotation_maint.py
Comment (mufernando) [inline maintenance/annotation_maint.py:1]: I tried to run this locally but got an error

```
~/projects/stdpopsim/maintenance annotation ?4 ❯ python3 annotation_maint.py                                                                                           stdpopsim-dev 13:25:30
Traceback (most recent call last):
  File "annotation_maint.py", line 21, in <module>
    x, y = urllib.request.urlretrieve(address, tmp_path)
  File "/opt/anaconda3/envs/stdpopsim-dev/lib/python3.7/urllib/request.py", line 288, in urlretrieve
    % (read, size), result)
urllib.error.ContentTooShortError: <urlopen error retrieval incomplete: got only 29639876 out of 37903989 bytes>

~/projects/stdpopsim/maintenance annotation ?5 ❯ python3 annotation_maint.py                                                                                   1m 48s  stdpopsim-dev 13:27:23
Traceback (most recent call last):
  File "annotation_maint.py", line 13, in <module>
    os.mkdir(annot_path)
FileExistsError: [Errno 17] File exists: 'annotations'
```

Comment (andrewkern) [reply maintenance/annotation_maint.py:1]: looks like there was a downloading error?

File: maintenance/__init__.py
Comment (andrewkern) [reply maintenance/__init__.py:4]: yeah, i'm not sure. i was trying to defeat circleci and ended up adding imports. will try to remove it.

File: requirements/development.txt
Comment (andrewkern) [reply requirements/development.txt:22]: yeah i can probably drop this. i was writing code to handle the AWS push for us, but because of user id issues this might not be feasible

File: stdpopsim/catalog/HomSap/__init__.py
Comment (andrewkern) [reply stdpopsim/catalog/HomSap/__init__.py:232]: looks like ensembl wants us to cite their most recent paper, if we are using their most recent releases: 

https://uswest.ensembl.org/info/about/publications.html

File: stdpopsim/catalog/HomSap/__init__.py
Comment (andrewkern) [reply stdpopsim/catalog/HomSap/__init__.py:224]: yup, this will be super easy. all one does is 

1. update the url string, 
2. rerun the maintenance code
3. upload to AWS

however the user will not be aware of an update if they have cached the annotations...

File: stdpopsim/species.py
Comment (mufernando) [inline stdpopsim/species.py:247]: i don't understand what this is doing

**Review by mufernando — COMMENTED** [review]
great work, Andy! just a couple of qs/comments

File: requirements/conda-minimal.txt
Comment (andrewkern) [reply requirements/conda-minimal.txt:9]: so i think circleci needs this here to run `test_maintence.py`

File: setup.py
Comment (andrewkern) [reply setup.py:36]: yeah again i added this for circleci... will clear these out

File: stdpopsim/annotations.py
Comment (andrewkern) [reply stdpopsim/annotations.py:17]: not sure what you mean here? do you mean filter on the maintenance side and then dump those into memory? these are btw quite small files-- human genome is 40Mb, so i'm not worried about memory.

File: stdpopsim/annotations.py
Comment (andrewkern) [reply stdpopsim/annotations.py:132]: not sure what you mean here? do you mean filter on the maintenance side and then dump those into memory? these are btw quite small files-- human genome is 40Mb, so i'm not worried about memory.

File: stdpopsim/species.py
Comment (andrewkern) [reply stdpopsim/species.py:247]: this is adding to the catalog of possible annotations in a Species

**Comment by andrewkern** [comment]
pushed the requested changes from @grahamgower and @mufernando. this _may_ break circleci ...

**Comment by andrewkern** [comment]
also worth mentioning that this now uses GRCh38 annotation from release 101 of Ensembl.

File: maintenance/annotation_maint.py
Comment (andrewkern) [reply maintenance/annotation_maint.py:13]: right so i like this because it fails if it's already there

File: stdpopsim/annotations.py
Comment (mufernando) [reply stdpopsim/annotations.py:132]: you could within `zarr_to_dataframe` do the filtering of rows on `seqid`.

Like

```
def zarr_to_dataframe(path, id):
    """
    converts zarr annotation file to
    pandas dataframe for manipulations
    """
    z = zarr.open(path)
   mask = z['seqid'][:] == id
    df = pandas.DataFrame(
        {'seqid': z['seqid'].get_mask_selection(mask),
         'source': z['source'].get_mask_selection(mask),
         'type': z['type'].get_mask_selection(mask),
         'start': z['start'].get_mask_selection(mask),
         'end': z['end'].get_mask_selection(mask),
         'score': z['score'].get_mask_selection(mask),
         'strand': z['strand'].get_mask_selection(mask),
         'phase': z['phase'].get_mask_selection(mask)}
    )
    return df


```

**Review by jradrion — APPROVED** [review]
LGTM, @andrewkern! The comments/requests from @grahamgower and @mufernando seem to have covered the relatively minor issues.

**Comment by jradrion** [comment]
just saw that there are new changes since I last checked out this PR, I'll take another look

**Review by jradrion — APPROVED** [review]
`test_get_genes_full (tests.test_annotations.TestGetChromosomeAnnotations) ... ok`
LGTM

File: stdpopsim/species.py
Comment (grahamgower) [inline stdpopsim/species.py:261]: So we'll need some cleverness added to `docs/_ext/speciescatalog.py` in order to get the annotation id's listed in the catalog docs. (just mentioning this so we remember to open a ticket when this is merged)

File: stdpopsim/catalog/HomSap/__init__.py
Comment (grahamgower) [inline stdpopsim/catalog/HomSap/__init__.py:227]: How about including the ensembl version in the zarr zip filename? Consider what happens if a user has an old version in their cache, and then the genome assembly is updated -> no new annotation file is downloaded.

File: stdpopsim/annotations.py
Comment (andrewkern) [reply stdpopsim/annotations.py:107]: yes this is a test for multiple threads downloading simultaneously

**Comment by andrewkern** [comment]
okay @grahamgower i think this is all now ready for merging. I've made all the requested changes-- thanks for the input @mufernando @jradrion @grahamgower

**Review by grahamgower — APPROVED** [review]
Great! Thanks @andrewkern! I'll leave it to you to merge, and open tickets for the remaining comments.

**Pull request merged**
